File: ts/module-config/push-to-var/fwk/cfg-src.ts
/*
 * Copyright © 2022 Paulo Villela. All rights reserved.
 * Use of this source code is governed by the MIT license
 * that can be found in the LICENSE file.
 */

import { makeMutableFunctionArity0, MutableFunctionArity0 } from "../../../arch/util/mutable-function-arity-0";

export type CfgSrc<T> = MutableFunctionArity0<T>

export function makeCfgSrc<T>(src: (() => T) | null): CfgSrc<T> {
  return makeMutableFunctionArity0(src)
}

type CfgSrcAdapter<S, T> = (s: S) => T

export class CfgSrcAdaptation<S, T> {
  constructor(
    private targetSrc: CfgSrc<T>,
    private adapter: CfgSrcAdapter<S, T>
  ) {
  }

  setOrigin(originSrc: () => S) {
    this.targetSrc.set(() => this.adapter(originSrc()))
  }
}

File: go/module-config/push-to-var/fwk/cfg-src.go
/*
 *  Copyright © 2021 Paulo Villela. All rights reserved.
 *  Use of this source code is governed by the Apache 2.0 license
 *  that can be found in the LICENSE file.
 */

package fwk

type CfgSrc[T any] func() T

func NilCfgSrc[T any]() T {
	panic("Module used before being initialized")
}

type CfgSrcAdapter[S, T any] func(S) T

type CfgSrcAdaptation[S, T any] struct {
	targetSrc *func() T
	adapter   CfgSrcAdapter[S, T]
}

func (s CfgSrcAdaptation[S, T]) SetOrigin(originSrc CfgSrc[S]) {
	*s.targetSrc = func() T { return s.adapter(originSrc()) }
}

func MakeCfgSrcAdaptation[S, T any](
	targetSrc *func() T,
	adapter CfgSrcAdapter[S, T],
) CfgSrcAdaptation[S, T] {
	return CfgSrcAdaptation[S, T]{
		targetSrc: targetSrc,
		adapter:   adapter,
	}
}
